### PR TITLE
Fixes #28927 - Forget host status

### PR DESCRIPTION
**Pull request by ianballou**
Adds the ability to remove an individual host status.  This is useful for old or unwanted warnings and errors.

At the moment it says "Forget Status", but there might be a better label, like "Clear Status" or something along those lines.  Perhaps even the clear button could be an icon, I'd be curious to hear what people have to say.  It'll make sense in the future for this action to be in the v2 API as well.

forget_status

### CODE REVIEW

**Comment by ares** [comment]
@ianballou this is awesome!

I'd consider shortening the link label and I like your suggestion with clearing. How about just "Clear"? Note it's already placed next to the specific host status, so we don't need to repeat it's clearing the status. Usually the status message is also not that long, so I'm adding another screenshot. @rohoover any recommendation from you?

forget

I tried to see, how it would look like with such short string and `float: right`

clear

such patch for the reference
```diff
diff --git a/app/helpers/host_description_helper.rb b/app/helpers/host_description_helper.rb
index 0cf163833..befca1c78 100644
--- a/app/helpers/host_description_helper.rb
+++ b/app/helpers/host_description_helper.rb
@@ -54,7 +54,7 @@ module HostDescriptionHelper
         _(status.name),
         content_tag(:span, ' '.html_safe, :class => host_global_status_icon_class(status.to_global)) +
           content_tag(:span, _(status.to_label), :class => host_global_status_class(status.to_global)) +
-          link_to(_(' Forget Status'), forget_status_host_path(host, status: status))
+          content_tag(:span, link_to(_('clear'), forget_status_host_path(host, status: status), :style => 'float:right'))
       ], :priority => priority += 1 }
     end.compact
   end
```

**Comment by timogoebel** [comment]
How about `Acknowledge` or `ACK`?

**Comment by ianballou** [comment]
I like how "clear" with `float: right` looks myself.  I think the intention is obvious enough given the context.

**Comment by ianballou** [comment]
It looks a little strange with longer messages (shown below), but I think the consistency of having "clear" in the bottom right corner is good.
katello_status

**Comment by ekohl** [comment]
Is there a nice (small) red button with icon styling that can be applied to indicate this is a destructive action? Other than that it makes sense to me.

As mentioned in person, I think the API can also be useful to clean up bulk actions but that doesn't block merging.

**Comment by ares** [comment]
Imho `Ack` wouldn't suggest it actually removes the information. Acknoledgement would be nice with also a timestamp "until", similarly like it is in Icinga. However this is not a downtime implementation. However I don't want to spend much time on the wording here, the sooner it gets in the better.

**Comment by ianballou** [comment]
[test katello]

File: app/helpers/host_description_helper.rb
Comment (ares) [inline app/helpers/host_description_helper.rb]: oops sorry, using `:style` directly was a quick hack to try out, I think it's better to use bootstrap CSS class called `float-right` so this line should be 

```
content_tag(:span, link_to(_('clear'), forget_status_host_path(host, status: status), :class => 'pull-right')),
```

**Review by ares — CHANGES_REQUESTED** [review]
I think it's quite close. One nit regarding style.

I also tried what @ekohl suggested - the icon, here's the screen
delete

for the reference, the code looked like this

```
content_tag(:span, link_to(icon_text('delete', '', kind: 'pficon', title: _('Delete')), forget_status_host_path(host, status: status), :class => 'pull-right')),
```

@Rohoover thoughts? Also @timogoebel and @ekohl which variant is acceptable for you? I still prefer `clear` text, since we don't use clickable icons elsewhere.

**Comment by ares** [comment]
Also @ianballou would you mind adding API endpoint for this? Would you need any help with such addition? FTR I'm fine with addressing that separately but at least the redmine issue should exist and should be set as a blocker for the next release.

**Comment by timogoebel** [comment]
> Also @timogoebel and @ekohl which variant is acceptable for you? I still prefer `clear` text, since we don't use clickable icons elsewhere.

Yep, clear is probably the best option.

**Comment by ekohl** [comment]
I agree that clear makes more sense in this context. Delete is confusing. To be clear (pun not intended), I was thinking of https://www.patternfly.org/v3/pattern-library/widgets/#buttons and in particular the danger styling.

Slight nitpick: currently it's a link (`<a>`) but since this action has a side effect, technically it should be a form so it ends up as a POST action rather than a GET. Browsers know not to repeat a POST when using the history while this is an issue with GET.

**Comment by ianballou** [comment]
@ares am I on the right track for creating the v2 API endpoint? I keep getting the error `unknown permission for api/v2/hosts#forget_status` even though I thought I took care of the permissions there.  I'm guessing once that's fixed I should add some error catching in the v2 hosts_controller method and an API test.

Also @ekohl I agree that POST makes more sense here.  I've started making the v2 endpoint POST as well.  Is there a helper I can use to make the clear link do a POST instead of a GET?

File: app/controllers/api/v2/hosts_controller.rb
Comment (ekohl) [inline app/controllers/api/v2/hosts_controller.rb]: Would a more restful way be `DELETE /hosts/:id/statuses/:status`

Comment (ianballou) [reply app/controllers/api/v2/hosts_controller.rb]: Actually I didn't realize it before but `/hosts/:id/status/:type` exists.  Perhaps I could do a `DELETE` to that?

**Comment by ianballou** [comment]
I updated the v2 API bits to use DELETE instead of POST.  Still seeing the permissions issue, hm.

File: config/routes.rb
Comment (timogoebel) [inline config/routes.rb]: I believe this should be
```suggestion
        delete 'forget_status'
```

**Review by timogoebel — COMMENTED** [review]
> I keep getting the error `unknown permission for api/v2/hosts#forget_status`

I believe you need to extend the `action_permission` method in the controller.

File: config/routes.rb
Comment (ianballou) [reply config/routes.rb]: Extending `action_permission` fixed it.  I do think it would make sense for the `get` there to be a `delete` or a `post`, but I'm not quite sure how to make the `clear` link not trigger a `get`.

**Comment by ianballou** [comment]
The v2 API endpoint is working now.  Let me know how that's looking. Still wondering about having the link trigger a POST or DELETE rather than GET.

File: config/routes.rb
Comment (timogoebel) [reply config/routes.rb]: @ianballou: Please find an example here (look for `method`): https://github.com/theforeman/foreman/blob/ae96cb70130968bdcc8b7a9a9af7edc7c6e0d9a8/app/views/key_pairs/index.html.erb#L28

File: app/controllers/api/v2/hosts_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/hosts_controller.rb]: Nit: In ruby you tyically use `==` for comparisons.
```suggestion
        if params[:type] == 'global'
```

File: app/controllers/api/v2/hosts_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/hosts_controller.rb]: If the status cannot be deleted, it would still show this success message.

**Comment by ianballou** [comment]
I think I addressed all the comments now.

**Review by ekohl — DISMISSED** [review]
With my limited rails knowledge this looks ok, but I'll leave this to the other reviewers to actually merge.

**Comment by ianballou** [comment]
@ares @timogoebel how's this PR looking?

File: app/controllers/api/v2/hosts_controller.rb
Comment (ares) [inline app/controllers/api/v2/hosts_controller.rb]: this could also return false without raising exception right? can we use `process_reponse` helper we use elsewhere? that would help with dedplucating the logic for false and for when exception is raised

I don't anticipate something would prevent status deletion and not causing exception, but it's probably better to make it robust from the beginning

File: app/controllers/api/v2/hosts_controller.rb
Comment (ares) [inline app/controllers/api/v2/hosts_controller.rb]: It actually shouldn't contain global right? It generates following in apipie doc

```
Must be one of: HostStatus::Global, configuration, build.
```

**Review by ares — CHANGES_REQUESTED** [review]
I promise this is the last thing. I tested and works well - both UI and API. The apipie documentation contains global which we don't support. Plus I left one ask related to code for the case deletion fails.

Happy to merge after that's resolved.

File: app/controllers/api/v2/hosts_controller.rb
Comment (ianballou) [reply app/controllers/api/v2/hosts_controller.rb]: I've updated the code to use `process_response`.  Is that the usage you were expecting? Wasn't sure if I should drop the rescue or not.

File: app/controllers/api/v2/hosts_controller.rb
Comment (ianballou) [reply app/controllers/api/v2/hosts_controller.rb]: Fixed!

**Review by ares — APPROVED** [review]
This is perfect! Retested and the API controller works like a charm. Apipie docs is now correct too.

**Comment by ares** [comment]
Thanks @ianballou, great contribution, please sign up for the community demo and show this great feature to the users! Merging!

File: app/controllers/api/v2/hosts_controller.rb
Comment (tbrisker) [inline app/controllers/api/v2/hosts_controller.rb:206]: @ianballou was there any reason you didn't use 
```suggestion
      param :type, HostStatus.status_registry.to_a.map { |s| s.humanized_name }, :required => true, :desc => N_("status type"))
```
here?

**Pull request merged**
